test/jdk/java/util/concurrent/StructuredTaskScope/StructuredTaskScopeTest.java

```diff
@@ -1019,247 +853,145 @@ void testInterruptClose2(ThreadFactory factory) throws Exception {
-            assertThrows(IllegalStateException.class, () -> {
-                scope.invokeEnsureOwnerAndJoined();
-            });
-
-            // owner thread, after join
-            scope.join();
-            scope.invokeEnsureOwnerAndJoined();
-
-            // thread in scope cannot invoke ensureOwnerAndJoined
-            Subtask<Boolean> subtask = scope.fork(() -> {
-                assertThrows(WrongThreadException.class, () -> {
-                    scope.invokeEnsureOwnerAndJoined();
-                });
-                return true;
-            });
-            scope.join();
-            assertTrue(subtask.get());
-
-            // random thread cannot invoke ensureOwnerAndJoined
-            try (var pool = Executors.newSingleThreadExecutor()) {
-                Future<Void> future = pool.submit(() -> {
-                    assertThrows(WrongThreadException.class, () -> {
-                        scope.invokeEnsureOwnerAndJoined();
-                    });
-                    return null;
-                });
-                future.get();
+            @Override
+            public Void result() {
+                return null;
             }
+        };
+        var excRef = new AtomicReference<Throwable>();
+        Thread.UncaughtExceptionHandler uhe = (t, e) -> excRef.set(e);
+        ThreadFactory factory = Thread.ofVirtual()
+                .uncaughtExceptionHandler(uhe)
+                .factory();
+        try (var scope = StructuredTaskScope.open(joiner, cf -> cf.withThreadFactory(factory))) {
+            scope.fork(() -> "foo");
+            scope.join();
+            assertInstanceOf(FooException.class, excRef.get());
         }
     }
 
     /**
-     * Test ensureOwnerAndJoined after the task scope has been closed.
+     * Test Joiner.onComplete returning true to cancel execution.
      */
-    @ParameterizedTest
-    @MethodSource("factories")
-    void testEnsureOwnerAndJoinedAfterClose(ThreadFactory factory) throws Exception {
-        class MyScope<T> extends StructuredTaskScope<T> {
-            MyScope(ThreadFactory factory) {
-                super(null, factory);
+    @Test
+    void testOnCompleteCancelsExecution() throws Exception {
+        var joiner = new Joiner<String, Void>() {
+            @Override
+            public boolean onComplete(Subtask<? extends String> subtask) {
+                return true;
             }
-            public void invokeEnsureOwnerAndJoined() {
-                super.ensureOwnerAndJoined();
+            @Override
+            public Void result() {
+                return null;
             }
-        }
-
-        // ensureOwnerAndJoined after close, join invoked
-        try (var scope = new MyScope<String>(factory)) {
+        };
+        try (var scope = StructuredTaskScope.open(joiner)) {
+            assertFalse(scope.isCancelled());
             scope.fork(() -> "foo");
+            while (!scope.isCancelled()) {
+                Thread.sleep(10);
```

Comment:
Might make sense to create a utility method to "awaitAndAssertCancel" since this location has a 10ms sleep but others have a 20ms sleep. (To ensure that it is handled uniformly)

Reply:
That's a good suggestion, an awaitCancelled would avoid having this while loop in a number of tests.